File: 2_Affine_cipher/Affine.py
"""
Affine with mod26 Cipher function to encrypt and decrypt 
"""
__author__ = "Vaasudevan (github.com/VaasuDevanS)"
__date__   = "Dec 20, 2018"


def affine_cipher(typ, sentence, a, b):

    crypted = []

    # Loop each character in sentence
    for s in sentence:
        if s.isalpha(): # [A-Za-Z]
            x = 65 if s.isupper() else 97
            if typ.lower().startswith("e"):
                crp_s = (a*(ord(s)-x) + b) % 26
            else:
                a_1 = {a*i%26 : i for i in range(26)}.get(1,0) # a inverse
                crp_s = a_1 * (ord(s)-x-b) % 26
            crypted.append(chr(crp_s+x))
        else:
            crypted.append(s)

    return "".join(crypted)

if __name__ == "__main__":

    E = affine_cipher("Encrypt", input("Enter the String: "), 5, 7)
    D = affine_cipher("Decrypt", E, 5, 7)

    print("Encrypted: %s\nDecrypted: %s" % (E, D))

# EOF

File: 2_Affine_cipher/Affine.cpp
/*
* Affine Cipher function based on ax+b and mod26.
* mod26 restricts the encrypted characters as alphabets
* Author: Vaasudevan (github.com/VaasuDevanS)
* Date: Dec 20, 2018
* Link: http://crypto.interactive-maths.com/affine-cipher.html
*/

#include <iostream>
#include <ctype.h>
using namespace std;

string encrypt(string in, int a, int b)
{
    string encrypted;
    for(char c : in) {
        // Encrypt if its only character [a-zA-Z]
        if ( (c>='a' && c<='z') || (c>='A' && c<='Z') ) {
            int flag = isupper(c) ? 65 : 97;
            int d = ( a * (c-flag) + b  ) % 26;
            encrypted.push_back(d + flag);
        }
        else {
            encrypted.push_back(c);
        }
    }
    return encrypted;
}

string decrypt(string enc, int a, int b)
{
    string decrypted; int a_1=0;
    // inverse
    for (int i=0; i<26; i++) {
        if (a*i%26 == 1)
            a_1 = i;
    }
    for(char c : enc) {
        // Decrypt if its only character[a-zA-Z]
        if ( (c>='a' && c<='z') || (c>='A' && c<='Z') ) {
            int flag = isupper(c) ? 65 : 97;
            int d = ( a_1 * (c-flag-b) ) % 26;
            d = (d<0)? d+26: d; // C performs modulus towards 0
            decrypted.push_back(d + flag);
        }
        else {
            decrypted.push_back(c);
        }
    }
    return decrypted;
}

int main()
{
    string input, encrypted, decrypted;
    cout << "Enter the String: ";
    getline(cin, input);

    encrypted = encrypt(input, 5, 7);     // Call Encrypt
    decrypted = decrypt(encrypted, 5, 7); // Decrypt the Encrypted

    // Print the Encrypted and the Decrypted
    cout << "Encrypted: " << encrypted << endl;
    cout << "Decrypted: " << decrypted << endl;
}

// EOF
